File: 10-heap/00215-Kth_Largest_Element_In_An_Array/00215-kth_largest_element_in_an_array.go
// 215: Kth Largest Element In An Array
// https://leetcode.com/problems/kth-largest-element-in-an-array

package main

import (
	"container/heap"
	"fmt"
)

type Heap []int

func (h Heap) Len() int            { return len(h) }
func (h Heap) Less(i, j int) bool  { return h[i] < h[j] }
func (h Heap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *Heap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *Heap) Pop() interface{} {
	n := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return n
}

// SOLUTION
func findKthLargest(nums []int, k int) int {
	pq := Heap(nums)
	heap.Init(&pq)

	for pq.Len() > k {
		heap.Pop(&pq)
	}

	return heap.Pop(&pq).(int)
}

func main() {
	// INPUT
	nums := []int{3, 2, 1, 5, 6, 4}
	k := 2

	// OUTPUT
	result := findKthLargest(nums, k)
	fmt.Println(result)
}

File: 10-heap/00215-Kth_Largest_Element_In_An_Array/00215-kth_largest_element_in_an_array.py
# 215: Kth Largest Element In An Array
# https://leetcode.com/problems/kth-largest-element-in-an-array


import heapq

class Solution:
    # SOLUTION
    def findKthLargest(self, nums: list[int], k: int) -> int:
        pq=[]
        count=0

        for num in nums:
            heapq.heappush(pq,num)
            count+=1
            if count > k:
                heapq.heappop(pq)
                count-=1

        return pq[0]


if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums = [3,2,1,5,6,4]
    k = 2

    # OUTPUT
    result = o.findKthLargest(nums, k)
    print(result)

